utils/getUser.ts

```diff
@@ -0,0 +1,19 @@
+import { getServerSession } from "next-auth";
+
+import { authOptions } from "../pages/api/auth/[...nextauth]";
+
+import type { User, Session } from "next-auth";
+
+export async function getUser() {
+  const session: Session = (await getServerSession(authOptions)) as Session;
+
+  if (!session) {
+    throw new Error("No session");
+  }
+  return session.user;
+}
+
+export async function getUsername() {
+  const user: User = await getUser();
```

Comment:
don't force-type this, it can hide errors (i.e. if it's not really a user that getUser() returns):
```suggestion
  const user = await getUser();
```

Reply:
This same concept is applicable to the rest of your code — try to delete all explicit typings. If something's not being inferred correctly, it's a hint that the types are diverging somewhere (which explicit types masks).